numba/np/random/generator_core.py

```diff
@@ -0,0 +1,121 @@
+"""
+Core Implementations for Generator/BitGenerator Models.
+"""
+
+from llvmlite import ir
+from numba.core import cgutils, types
+from numba.core.extending import (intrinsic, make_attribute_wrapper, models,
+                                  overload, register_jitable,
+                                  register_model)
+from numba.np.numpy_support import numpy_version
+
+
+@register_model(types.NumPyRandomBitGeneratorType)
+class NumPyRngBitGeneratorModel(models.StructModel):
+    def __init__(self, dmm, fe_type):
+        members = [
+            ('parent', types.pyobject),
+            ('state_address', types.uintp),
+            ('state', types.uintp),
+            ('fnptr_next_uint64', types.uintp),
+            ('fnptr_next_uint32', types.uintp),
+            ('fnptr_next_double', types.uintp),
+            ('bit_generator', types.uintp),
+        ]
+        super(NumPyRngBitGeneratorModel, self).__init__(dmm, fe_type, members)
+
+
+_bit_gen_type = types.NumPyRandomBitGeneratorType('bit_generator')
+
+
+@register_model(types.NumPyRandomGeneratorType)
+class NumPyRandomGeneratorTypeModel(models.StructModel):
+    def __init__(self, dmm, fe_type):
+        members = [
+            ('bit_generator', _bit_gen_type),
+            ('meminfo', types.MemInfoPointer(types.voidptr)),
+            ('parent', types.pyobject)
+        ]
+        super(
+            NumPyRandomGeneratorTypeModel,
+            self).__init__(
+            dmm,
+            fe_type,
+            members)
+
+
+# The Generator instances have a bit_generator attr
+make_attribute_wrapper(
+    types.NumPyRandomGeneratorType,
+    'bit_generator',
+    'bit_generator')
+
+
+# Generate the overloads for "next_(some type)" functions
+def _generate_next_binding(overloadable_function, return_type):
```

Comment:
This looks like it would make for a better docstring than comment.  At the very least, as a docstring, this information will be more readily available during interactive debugging.

Reply:
Will add that in #8031 and rebase this PR upon it. (For context, we had to split the PRs due to their size and the kind of functionality it adds.)